src/libcore/ptr.rs

```diff
@@ -228,6 +228,67 @@ pub const fn null<T>() -> *const T { 0 as *const T }
 #[rustc_promotable]
 pub const fn null_mut<T>() -> *mut T { 0 as *mut T }
 
+#[repr(C)]
+union Repr<T> {
+    rust: *const [T],
+    rust_mut: *mut [T],
+    raw: FatPtr<T>,
+}
+
+#[repr(C)]
+struct FatPtr<T> {
+    data: *const T,
+    len: usize,
+}
+
+/// Forms a slice from a pointer and a length.
+///
+/// The `len` argument is the number of **elements**, not the number of bytes.
+///
+/// # Panics
+///
+/// The total size of the slice should be no
+/// larger than `isize::MAX` **bytes** in memory.
+///
+/// # Examples
+///
+/// ```rust
+/// #![feature(slice_from_raw_parts)]
+/// use std::ptr;
+///
+/// // create a slice pointer when starting out with a pointer to the first element
+/// let mut x = [5, 6, 7];
+/// let ptr = &mut x[0] as *mut _;
+/// let slice = ptr::slice_from_raw_parts_mut(ptr, 3);
+/// assert_eq!(unsafe { &*slice }[2], 7);
+/// ```
+#[inline]
+#[unstable(feature = "slice_from_raw_parts", reason = "recently added", issue = "36925")]
+pub fn slice_from_raw_parts<T>(data: *const T, len: usize) -> *const [T] {
+    debug_assert!(mem::size_of::<T>().saturating_mul(len) <= isize::max_value() as usize,
+                  "attempt to create slice covering half the address space");
+    unsafe { Repr { raw: FatPtr { data, len } }.rust }
+}
+
+/// Performs the same functionality as [`from_raw_parts`], except that a
+/// mutable slice is returned.
+///
+/// # Panics
+///
+/// The total size of the slice should be no
+/// larger than `isize::MAX` **bytes** in memory.
+///
+/// See the documentation of [`from_raw_parts`] for more details.
+///
+/// [`from_raw_parts`]: ../../std/slice/fn.from_raw_parts.html
+#[inline]
+#[unstable(feature = "slice_from_raw_parts", reason = "recently added", issue = "36925")]
+pub fn slice_from_raw_parts_mut<T>(data: *mut T, len: usize) -> *mut [T] {
+    debug_assert!(mem::size_of::<T>().saturating_mul(len) <= isize::max_value() as usize,
+                  "attempt to create slice covering half the address space");
```

Comment:
Is this actually a safety issue anywhere? Specifically, is it always safe to create a pointer to a slice with a length greater than `isize::MAX` if you don't dereference it?

IIUC due to how `libcore` is compiled this function won't actually panic in user code, which could be confusing.

Reply:
It's not a safety issue. I mostly thought of it as "you're going into weird places"-territory. There's basically no reason to ever create such raw slices. I should remove the `Panics` note though. This is just a debug assertion that we'd only get on the builders with debug assertions, so it's mostly a sanity test for our own test suite.

Reply:
*oops* seems like I had the same question.^^ But I don't think it is immediately obvious that there is no safety issue here. There's a decision being made here, which we cannot take back once we stabilize this.